### PR TITLE
kubelet: Make service environment variables optional

### LINKED ISSUE #60099: Service-environment variables should be optional
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:
Every container gets ~8 environmentvariables per service exposing a lot of information, and there is no way to turn it off.

**What you expected to happen**:
This `docker` deprecated `--link` feature compatibility option to be an opt-in (or at least opt-out) option.

**How to reproduce it (as minimally and precisely as possible)**:
Create a new container in a cluster where there are existing services. Check the `env`. You should see something like:

```
/ # env | grep KUBERNETES
KUBERNETES_SERVICE_PORT=443
KUBERNETES_PORT=tcp://10.233.0.1:443
KUBERNETES_PORT_443_TCP_ADDR=10.233.0.1
KUBERNETES_PORT_443_TCP_PORT=443
KUBERNETES_PORT_443_TCP_PROTO=tcp
KUBERNETES_PORT_443_TCP=tcp://10.233.0.1:443
KUBERNETES_SERVICE_PORT_HTTPS=443
KUBERNETES_SERVICE_HOST=10.233.0.1

# one ingress-controller
/ # env | grep INGRESS_NGINX_EXT | wc -l
36
```

**Anything else we need to know?**:
The code doing this is https://github.com/kubernetes/kubernetes/blob/v1.9.3/pkg/kubelet/kubelet_pods.go#L551-L553, and some docs at https://kubernetes.io/docs/concepts/containers/container-environment-variables/.

Maybe things like this should be exposed using eg the downwards api instead (https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/)

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-09T21:51:06Z", GoVersion:"go1.9.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2+coreos.0", GitCommit:"b427929b2982726eeb64e985bc1ebb41aaa5e095", GitTreeState:"clean", BuildDate:"2018-01-18T22:56:14Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration: on-prem
- OS (e.g. from /etc/os-release): ubuntu
- Kernel (e.g. `uname -a`):
- Install tools: helm
- Others:

### PR DESCRIPTION
**Pull request by vadorovsky**
**What this PR does / why we need it**:
Services should be accessed by DNS, not by environment variables which are associated with the deprecated --link option in Docker. In this change, those variables can be optionally enabled by setting up the `kubernetes.io/service-environment-variables` annotation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #60099

**Special notes for your reviewer**:

**Release note**:

```release-note
Make service environment variables optional
```

### CODE REVIEW

**Comment by thockin** [comment]
Please start with this:  https://github.com/kubernetes/community/pull/1249

We have a mostly-agreed design.

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -557,18 +558,26 @@ func (kl *Kubelet) getServiceEnvVarMap(ns string) (map[string]string, error) {
...
+	// DNS should be used instead of env vars - which is the default behavior.
+	//
 	// Note that there is a race between Kubelet seeing the pod and kubelet seeing the service.
 	// To avoid this users can: (1) wait between starting a service and starting; or (2) detect
 	// missing service env var and exit and be restarted; or (3) use DNS instead of env vars
 	// and keep trying to resolve the DNS name of the service (recommended).
-	serviceEnv, err := kl.getServiceEnvVarMap(pod.Namespace)
-	if err != nil {
-		return result, err
+	if _, ok := pod.Annotations[kubeapi.ServiceEnvironmentVariablesKey]; ok {
```
Comment (thockin) [inline pkg/kubelet/kubelet_pods.go]: This should be a field, not an annotation.

Comment (mrostecki) [reply pkg/kubelet/kubelet_pods.go]: done

**Comment by thockin** [comment]
I very much want to see this happen.  I hope you have time to follow through.  Please assign to me when ready for review.

**Comment by bradhoekstra** [comment]
I'm also interested in this feature. It looks like it hasn't been touched in while.

@mrostecki I am willing to help out/take over completion of this PR if you aren't working on it right now.

**Comment by mrostecki** [comment]
@bradhoekstra Sorry for the lack of activity. I will rebase and fix this PR today, in regard to the latest discussion on the proposal.

**Comment by bradhoekstra** [comment]
Perfect, thanks!

**Comment by thockin** [comment]
please assign me when ready.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -2954,6 +2954,10 @@ type PodSpec struct {
 	// configuration based on DNSPolicy.
 	// +optional
 	DNSConfig *PodDNSConfig `json:"dnsConfig,omitempty" protobuf:"bytes,26,opt,name=dnsConfig"`
+	// EnableServiceLinks indicates whether information about services should be injected into pod's
```
Comment (thockin) [inline staging/src/k8s.io/api/core/v1/types.go]: document the default

File: pkg/kubelet/kubelet_pods.go [hunk 3]
```diff
@@ -538,6 +538,11 @@ func (kl *Kubelet) getServiceEnvVarMap(ns string) (map[string]string, error) {
 		//
 		// ordering of the case clauses below enforces this
 		case ns:
+			// If service links are not enabled, we shouldn't add services from
+			// the same namespace.
+			if !*enableServiceLinks {
```
Comment (thockin) [inline pkg/kubelet/kubelet_pods.go]: I think we need to special-case KUBERNETES_* vars (master) because clients need it.

**Comment by bradhoekstra** [comment]
@mrostecki Any progress on this?

**Comment by mrostecki** [comment]
@bradhoekstra Yes, please expect update on my PR tomorrow. Sorry for making you wait!

File: pkg/kubelet/kubelet_pods.go [hunk 4]
```diff
@@ -536,6 +536,11 @@ func (kl *Kubelet) getServiceEnvVarMap(ns string) (map[string]string, error) {
 	}
 
 	for _, e := range envvars.FromServices(mappedServices) {
+		// If service links are not enabled, we shouldn't add service variables
+		// except those which start from "KUBERNETES_".
+		if !*enableServiceLinks && !strings.HasPrefix(e.Name, "KUBERNETES_") {
```
Comment (thockin) [inline pkg/kubelet/kubelet_pods.go]: It's rather wasteful to call FromServices and then throw away most of it, especially if we want this to be the default mode eventually.

Can we instead change the loop above to test the flag before adding services to the `serviceMap` ?

File: pkg/kubelet/kubelet_pods.go [hunk 5]
```diff
@@ -544,6 +549,8 @@ func (kl *Kubelet) getServiceEnvVarMap(ns string) (map[string]string, error) {
 // Make the environment variables for a pod in the given namespace.
 func (kl *Kubelet) makeEnvironmentVariables(pod *v1.Pod, container *v1.Container, podIP string) ([]kubecontainer.EnvVar, error) {
 	var result []kubecontainer.EnvVar
+	var serviceEnv map[string]string
```
Comment (thockin) [inline pkg/kubelet/kubelet_pods.go]: Why this change?  In general, scoping closer to usage is better.

File: pkg/kubelet/kubelet_pods_test.go [hunk 6]
```diff
@@ -834,20 +834,22 @@ func TestMakeEnvironmentVariables(t *testing.T) {
...
+		expectedEnvs       []kubecontainer.EnvVar // a set of expected environment vars
+		expectedError      bool                   // does the test fail
+		expectedEvent      string                 // does the test emit an event
 	}{
 		{
-			name: "api server = Y, kubelet = Y",
-			ns:   "test1",
+			name:               "api server = Y, kubelet = Y",
+			ns:                 "test1",
+			enableServiceLinks: false,
```
Comment (thockin) [inline pkg/kubelet/kubelet_pods_test.go]: It's not clear what these test cases think they are testing, but you're changing it.  Since the default is true, you should probably handle this as:

1) add `enableServiceLinks: true` to each test case
2) add new test cases (with comments, please) that explicitly exercise the `false` case of that and nothing else.

If you feel generous you could refactor this test, making a util function that takes a list of test cases and a bunch of smaller Test* functions that call that util function with more targetted cases, e.g `TestMakeEnvironmentVariablesWithExpansions` and `TestMakeEnvironmentVariablesFromConfigMap` etc.  You can do that a a pre-factoring commit in this same PR.  If you feel generous...  It would make this testing much nicer :)

**Comment by mrostecki** [comment]
@bradhoekstra OK, we can close this one in favor of yours.
